File: app/src/main/java/com/o4x/musical/ui/activities/details/AbsDetailActivity.java
package com.o4x.musical.ui.activities.details;

import android.content.Intent;
import android.content.res.ColorStateList;
import android.os.Bundle;
import android.text.Spanned;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.core.widget.NestedScrollView;
import androidx.loader.app.LoaderManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.afollestad.materialcab.MaterialCab;
import com.afollestad.materialdialogs.MaterialDialog;
import com.afollestad.materialdialogs.util.DialogUtils;
import com.google.android.material.textview.MaterialTextView;
import com.o4x.musical.R;
import com.o4x.musical.imageloader.universalil.listener.PaletteMusicLoadingListener;
import com.o4x.musical.imageloader.universalil.loader.UniversalIL;
import com.o4x.musical.interfaces.CabHolder;
import com.o4x.musical.interfaces.PaletteColorHolder;
import com.o4x.musical.model.Song;
import com.o4x.musical.ui.activities.base.AbsMusicPanelActivity;
import com.o4x.musical.ui.adapter.song.DetailsSongAdapter;
import com.o4x.musical.util.PhonographColorUtil;
import com.o4x.musical.util.Util;
import com.o4x.musical.util.color.MediaNotificationProcessor;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import code.name.monkey.appthemehelper.util.ColorUtil;
import code.name.monkey.appthemehelper.util.ToolbarContentTintHelper;

public abstract class AbsDetailActivity<T> extends AbsMusicPanelActivity implements PaletteColorHolder, CabHolder, LoaderManager.LoaderCallbacks<T> {

    static final int TAG_EDITOR_REQUEST = 2001;

    @BindView(R.id.nested_scroll_view)
    NestedScrollView scrollView;
    @BindView(R.id.song_recycler)
    RecyclerView songRecyclerView;
    @BindView(R.id.image)
    ImageView image;
    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.header)
    View headerView;
    @BindView(R.id.gradient)
    View gradient;
    @BindView(R.id.title)
    MaterialTextView title;
    @BindView(R.id.subtitle)
    MaterialTextView subtitle;

    MaterialCab cab;
    int imageHeight;
    MediaNotificationProcessor colors;

    @Nullable
    Spanned wiki;
    MaterialDialog wikiDialog;

    DetailsSongAdapter songAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ButterKnife.bind(this);

        setDrawUnderStatusBar();
        setUpToolBar();
        setupViews();

        reload();
    }

    @Override
    protected View createContentView() {
        return wrapSlidingMusicPanel(R.layout.activity_detail);
    }

    @Override
    public int getPaletteColor() {
        return colors.getActionBarColor();
    }

    @NonNull
    @Override
    public MaterialCab openCab(int menuRes, @NonNull final MaterialCab.Callback callback) {
        if (cab != null && cab.isActive()) cab.finish();
        cab = new MaterialCab(this, R.id.cab_stub)
                .setMenu(menuRes)
                .setCloseDrawableRes(R.drawable.ic_close_white_24dp)
                .setBackgroundColor(PhonographColorUtil.shiftBackgroundColorForLightText(getPaletteColor()))
                .start(new MaterialCab.Callback() {
                    @Override
                    public boolean onCabCreated(MaterialCab materialCab, Menu menu) {
                        return callback.onCabCreated(materialCab, menu);
                    }

                    @Override
                    public boolean onCabItemClicked(MenuItem menuItem) {
                        return callback.onCabItemClicked(menuItem);
                    }

                    @Override
                    public boolean onCabFinished(MaterialCab materialCab) {
                        return callback.onCabFinished(materialCab);
                    }
                });
        return cab;
    }

    @Override
    public void onBackPressed() {
        if (cab != null && cab.isActive()) cab.finish();
        else {
            scrollView.stopNestedScroll();
            super.onBackPressed();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == TAG_EDITOR_REQUEST) {
            reload();
            setResult(RESULT_OK);
        }
    }

    private void setUpToolBar() {
        setSupportActionBar(toolbar);
        //noinspection ConstantConditions
        getSupportActionBar().setTitle(null);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    }

    void setupViews() {
        setupScrollView();
        setupSongsRecycler();
        setColors(DialogUtils.resolveColor(this, R.attr.defaultFooterColor), null);
    }

    void setupScrollView() {
        final int displayHeight = Util.getScreenHeight();
        imageHeight = Util.getScreenWidth();
        final int gradientHeight =
                (int) (imageHeight + getResources().getDimension(R.dimen.detail_header_height));

        scrollView.setOnScrollChangeListener(new NestedScrollView.OnScrollChangeListener() {
            @Override
            public void onScrollChange(NestedScrollView v, int scrollX, int scrollY, int oldScrollX, int oldScrollY) {

                // Change alpha of overlay
                final float alpha = Math.max(0, Math.min(1, (float) 2 * scrollY / gradientHeight));
                setAppbarAlpha(alpha);

                // Scroll poster
                image.setTranslationY(
                        Math.max(-scrollY / (displayHeight * 2 / imageHeight), -imageHeight)
                );
            }
        });
    }

    void setupSongsRecycler() {
        songAdapter = new DetailsSongAdapter(this, getSongs(), R.layout.item_list, false, this);
        songRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        songRecyclerView.setAdapter(songAdapter);
        songAdapter.registerAdapterDataObserver(new RecyclerView.AdapterDataObserver() {
            @Override
            public void onChanged() {
                super.onChanged();
                if (songAdapter.getItemCount() == 0) finish();
            }
        });
    }

    private void setColors(int color, @Nullable MediaNotificationProcessor colors) {
        if (colors != null) {
            this.colors = colors;
            ToolbarContentTintHelper.colorizeToolbar(toolbar, colors.getPrimaryTextColor(), this);
            setNavigationBarColor(colors.getActionBarColor());
            setTaskDescriptionColor(colors.getActionBarColor());

            title.setTextColor(colors.getPrimaryTextColor());
            subtitle.setTextColor(colors.getSecondaryTextColor());
        }

        gradient.setBackgroundTintList(ColorStateList.valueOf(color));
        headerView.setBackgroundColor(color);
        findViewById(android.R.id.content).getRootView().setBackgroundColor(color);

        setAppbarAlpha(0);
    }

    private void setAppbarAlpha(float alpha) {
        if (colors == null) return;
        toolbar.setBackgroundColor(
                ColorUtil.INSTANCE.withAlpha(colors.getActionBarColor(), alpha)
        );
        setStatusBarColor(
                ColorUtil.INSTANCE.withAlpha(colors.getActionBarColor(), alpha)
        );
        toolbar.setTitleTextColor(
                ColorUtil.INSTANCE.withAlpha(colors.getPrimaryTextColor(), alpha)
        );
    }

    @Override
    public void onMediaStoreChanged() {
        super.onMediaStoreChanged();
        reload();
    }

    UniversalIL getImageLoader() {
        return new UniversalIL(
                image,
                new PaletteMusicLoadingListener() {
                    @Override
                    public void onColorReady(@NotNull MediaNotificationProcessor colors) {
                        setColors(colors.getBackgroundColor(), colors);
                        setMiniPlayerColor(colors);
                        songAdapter.setColors(colors);
                    }
                }, imageHeight
        );
    }

    abstract void loadImage();
    abstract void reload();
    protected abstract List<Song> getSongs();
}

File: app/src/main/java/com/o4x/musical/ui/activities/details/AbsDetailActivity.kt
package com.o4x.musical.ui.activities.details

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.AdapterDataObserver
import com.o4x.appthemehelper.extensions.surfaceColor
import com.o4x.appthemehelper.extensions.textColorPrimary
import com.o4x.appthemehelper.extensions.textColorSecondary
import com.o4x.appthemehelper.extensions.withAlpha
import com.o4x.appthemehelper.util.ColorUtil.withAlpha
import com.o4x.appthemehelper.util.ToolbarContentTintHelper
import com.afollestad.materialcab.attached.AttachedCab
import com.afollestad.materialcab.attached.destroy
import com.afollestad.materialcab.attached.isActive
import com.afollestad.materialcab.createCab
import com.o4x.musical.App
import com.o4x.musical.R
import com.o4x.musical.imageloader.glide.loader.GlideLoader
import com.o4x.musical.imageloader.glide.targets.palette.NotificationPaletteTargetListener
import com.o4x.musical.interfaces.CabCallback
import com.o4x.musical.model.Song
import com.o4x.musical.prefs.PreferenceUtil
import com.o4x.musical.ui.activities.base.AbsMusicPanelActivity
import com.o4x.musical.ui.adapter.song.DetailsSongAdapter
import com.o4x.musical.ui.viewmodel.ScrollPositionViewModel
import com.o4x.musical.util.Util
import com.o4x.musical.util.ViewUtil
import com.o4x.musical.util.color.MediaNotificationProcessor
import kotlinx.android.synthetic.main.activity_detail.*
import org.koin.androidx.viewmodel.ext.android.viewModel
import kotlin.math.max
import kotlin.math.min

abstract class AbsDetailActivity<T> : AbsMusicPanelActivity() {

    companion object {
        const val EXTRA_ID = "extra_id"
        const val TAG_EDITOR_REQUEST = 2001
    }

    val scrollPositionViewModel by viewModel<ScrollPositionViewModel>()

    var data: T? = null


    var imageHeight: Int? = null
    var displayHeight: Int? = null
    var gradientHeight: Int? = null
    lateinit var colors: MediaNotificationProcessor

    var songAdapter: DetailsSongAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setDrawUnderStatusBar()

        setupImage()
        initObserver()
        setUpToolBar()
        setupViews()
    }

    override fun createContentView(): View? {
        return wrapSlidingMusicPanel(R.layout.activity_detail)
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == TAG_EDITOR_REQUEST) {
            initObserver()
            setResult(RESULT_OK)
        }
    }

    private fun setupImage() {
        imageHeight = Util.getScreenWidth()
        colors = MediaNotificationProcessor(this)
        displayHeight = Util.getScreenHeight()
        gradientHeight =
            (imageHeight!! + resources.getDimension(R.dimen.detail_header_height)).toInt()
    }

    private fun setUpToolBar() {
        setSupportActionBar(toolbar)
        supportActionBar?.title = null
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        toolbar.post {
            ToolbarContentTintHelper.colorizeToolbar(toolbar, colors.primaryTextColor, this)
            setAppbarAlpha(0f)
        }
    }

    open fun setupViews() {
        setupSongsRecycler()
        setupScrollView()
    }

    private fun setupSongsRecycler() {
        songAdapter = DetailsSongAdapter(
            this, getSongs(), R.layout.item_list,  this, data!!, colors
        )
        song_recycler.layoutManager = LinearLayoutManager(this)
        song_recycler.adapter = songAdapter
        songAdapter?.registerAdapterDataObserver(object : AdapterDataObserver() {
            override fun onChanged() {
                super.onChanged()
                if (songAdapter!!.itemCount == 0) finish()
            }
        })
    }


    private fun setupScrollView() {
        song_recycler.post {
            song_recycler.layoutManager?.scrollToPosition(0)
            scrollPositionViewModel.setPosition(0)
            image.translationY = 0f
            setAppbarAlpha(0f)
        }

        song_recycler.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)

                scrollPositionViewModel.addPosition(dy)
                val scrollY = scrollPositionViewModel.getPositionValue()

                onScrollChange(scrollY)
            }
        })
    }

    fun onScrollChange(scrollY: Int) {
        if (cab?.isActive() != true) {
            // Change alpha of overlay
            val alpha = max(0f, min(0.9f, 2f * scrollY / gradientHeight!!))
            setAppbarAlpha(alpha)
        }

        // Scroll poster
        image.translationY =
            max(-scrollY / (displayHeight!! * 2 / imageHeight!!), -imageHeight!!)
                .toFloat()
    }

    override fun openCab(menuRes: Int, callback: CabCallback): AttachedCab {
        if (cab != null && cab!!.isActive()) cab!!.destroy()

        cab = createCab(R.id.cab_stub) {
            closeDrawable(R.drawable.ic_close)
            backgroundColor(literal = colors.backgroundColor)
            titleColor(literal = colors.primaryTextColor)
            subtitleColor(literal = colors.secondaryTextColor)
            popupTheme(PreferenceUtil.getGeneralThemeRes())
            menu(menuRes)
            menuIconColor(literal = colors.primaryTextColor)

            onCreate { cab, menu ->
                callback.onCreate(cab, menu)
                setAppbarAlpha(1f)
            }
            onSelection { item ->
                return@onSelection callback.onSelection(item)
            }
            onDestroy { cab ->
                onScrollChange(scrollPositionViewModel.getPositionValue())
                return@onDestroy callback.onDestroy(cab)
            }
        }

        return cab!!
    }


    fun setAllColors(colors: MediaNotificationProcessor) {
        this.colors = colors
        songAdapter?.colors = colors

        ToolbarContentTintHelper.colorizeToolbar(toolbar, colors.primaryTextColor, this)
        setNavigationBarColor(colors.backgroundColor)
        setTaskDescriptionColor(colors.backgroundColor)

        ViewUtil.setScrollBarColor(song_recycler, colors.secondaryTextColor.withAlpha(.3f))

        findViewById<View>(android.R.id.content).rootView.setBackgroundColor(colors.backgroundColor)
    }

    private fun setAppbarAlpha(alpha: Float) {
        toolbar.setBackgroundColor(
            withAlpha(colors.backgroundColor, alpha)
        )
        setStatusBarColor(
            withAlpha(colors.backgroundColor, alpha)
        )
    }

    fun getImageLoader(): GlideLoader.GlideBuilder {
        // if we use this for context glide on load sync in rotation will crashed
        return GlideLoader.with(App.getContext())
            .withListener(object : NotificationPaletteTargetListener(this) {
                override fun onColorReady(colors: MediaNotificationProcessor) {
                    setAllColors(colors)
                    setMiniPlayerColor(colors)
                }
            }.apply { loadPlaceholderPalette = true })
    }

    abstract fun initObserver()
    abstract fun loadImage()
    protected abstract fun getSongs(): List<Song>
}